PostBy/View Models/PostViewModel.m

```diff
@@ -125,29 +125,103 @@ - (void) checkIfDisliked {
 }
 
 - (void) likeButtonTap {
-    if (self.isLiked) {
-        [self unlikePost];
-    } else {
-        [self likePost];
-        if (self.isDisliked) {
-            [self undislikePost];
-        }
-    }
-    
-    [self saveAndRefresh];
+    [self interactWithPostBy:@"like"];
 }
 
 - (void) dislikeButtonTap {
-    if (self.isDisliked) {
-        [self undislikePost];
-    } else {
-        [self dislikePost];
-        if (self.isLiked) {
-            [self unlikePost];
+    [self interactWithPostBy:@"dislike"];
+}
+
+- (void) interactWithPostBy:(NSString *)action {
+    // construct query
+    PFQuery *query = [PFQuery queryWithClassName:@"Post"];
+    [query includeKeys:@[@"author"]];
+    [query whereKey:@"objectId" equalTo:self.post.objectId];
+
+    // fetch data asynchronously
+    [query findObjectsInBackgroundWithBlock:^(NSArray *posts, NSError *error) {
+        if (posts != nil) {
+            if (posts.count > 0) {
+                // Update queried post to reflect local post
+                Post *queriedPost = posts[0];
+                [self resetParsePostToLocalPost:queriedPost];
+                
+                // Update our post to be the queried post without the like/dislike as DB
+                self.post = queriedPost;
+                
+                [PFUser.currentUser save];
+                [self.post save];
+                
+                // Perform action based on local post
+                if ([action isEqualToString:@"like"]) {
+                    if (self.isLiked) {
+                        [self unlikePost];
+                    } else {
+                        [self likePost];
+                        if (self.isDisliked) {
+                            [self undislikePost];
+                        }
+                    }
+                } else if ([action isEqualToString:@"dislike"]) {
+                    if (self.isDisliked) {
+                        [self undislikePost];
+                    } else {
+                        [self dislikePost];
+                        if (self.isLiked) {
+                            [self unlikePost];
+                        }
+                    }
+                }
+                [self saveAndRefresh];
+            } else {
+                // Invalid / Deleted Post
+                [self.delegate postNotFound:self];
+            }
+        } else {
+            NSLog(@"Error: %@", error.localizedDescription);
         }
+    }];
+}
+
+- (void) resetParsePostToLocalPost:(Post *)queriedPost {
```

Comment:
The same for functionality decomposition.

It is better to consider decomposing the function and wrapping like/dislike logic into another function call(s).

Reply:
Good point Rao! I fixed that in one of the newest commits. I will look into doing in-depth testing on these like/dislike functions on Sprint 4 in order to allow me to complete other features.